src/process/receipts.rs

```diff
@@ -0,0 +1,297 @@
+    let process_receipt_actions_future = process_receipt_actions(&pool, &receipts_to_process);
+
+    let process_receipt_data_future = process_receipt_data(&pool, &receipts_to_process);
+
+    join!(
+        save_receipts_future,
+        process_receipt_actions_future,
+        process_receipt_data_future
+    );
+}
+
+async fn save_receipts(
+    pool: &Pool<ConnectionManager<PgConnection>>,
+    receipts: Vec<models::Receipt>,
+) {
+    loop {
+        match diesel::insert_into(schema::receipts::table)
+            .values(receipts.clone())
+            .execute_async(&pool)
+            .await
+        {
+            Ok(_) => break,
+            Err(async_error) => {
+                error!(
+                    target: crate::INDEXER_FOR_EXPLORER,
+                    "Error occurred while Receipt were adding to database. Retrying in {} milliseconds... \n {:#?} \n{:#?}",
+                    crate::INTERVAL.as_millis(),
+                    async_error,
+                    receipts,
+                );
+                tokio::time::delay_for(crate::INTERVAL).await;
+            }
+        };
+    }
+}
+
+async fn process_receipt_actions(
+    pool: &Pool<ConnectionManager<PgConnection>>,
+    receipts: &[&near_indexer::near_primitives::views::ReceiptView],
+) {
+    let mut receipt_actions: Vec<models::ReceiptAction> = vec![];
+    let mut receipt_action_actions: Vec<models::ReceiptActionAction> = vec![];
+    let mut receipt_action_input_data: Vec<models::ReceiptActionInputData> = vec![];
+    let mut receipt_action_output_data: Vec<models::ReceiptActionOutputData> = vec![];
+
+    for receipt in receipts {
+        if let near_indexer::near_primitives::views::ReceiptEnumView::Action {
+            actions,
+            input_data_ids,
+            output_data_receivers,
+            ..
+        } = &receipt.receipt
+        {
+            if let Ok(model) = models::ReceiptAction::try_from(*receipt) {
+                receipt_actions.push(model);
+
+                for (index, action) in actions.iter().enumerate() {
+                    receipt_action_input_data.extend(input_data_ids.iter().map(|data_id| {
+                        models::ReceiptActionInputData::from_data_id(
+                            receipt.receipt_id.to_string(),
+                            data_id.to_string(),
+                        )
+                    }));
+                    receipt_action_output_data.extend(output_data_receivers.iter().map(
+                        |receiver| {
+                            models::ReceiptActionOutputData::from_data_receiver(
+                                receipt.receipt_id.to_string(),
+                                receiver,
+                            )
+                        },
+                    ));
+                    receipt_action_actions.push(models::ReceiptActionAction::from_action_view(
+                        receipt.receipt_id.to_string(),
+                        i32::from_usize(index).unwrap(),
```

Comment:
I think it is fine to use `as i32` with a comment that we don't expect to go over i32::MAX of actions in a single receipt. Ideally, I would prefer to have `enumerate` produce i32 values, but we don't have such an option. To be on a safer side, you may choose to use `.expect()` with a clear message that we never expected to overflow here.